File: node/deliberation/src/specialists/defi.mjs
/**
 * @accretive-substrate/deliberation/specialists/defi
 *
 * DeFi specialist. Filters venue in {aerodrome, uniswap}; reads pool APR
 * and arbitrage spread state.
 *
 * Verdict rules:
 *   pool_apr > 50% + tvl > $1M               → buy   (high-yield viable pool)
 *   arb_spread > 1.5%                         → buy   (arb opportunity)
 *   pool_apr < 5% + no_arb                    → hold  (no edge)
 *   impermanent_loss_pct > 8% (open position) → sell  (IL eating yield)
 *   else                                      → hold
 *
 * Domain: 'defi'.
 */

import { makeSpecialist, voteFor, abstainBecause } from './base.mjs';

export const DEFI_SPECIALIST_NAME = 'defi';

async function argue(pkg) {
  const venue = String(pkg?.trigger?.venue || '').toLowerCase();
  if (!['aerodrome', 'uniswap'].includes(venue)) {
    return abstainBecause(DEFI_SPECIALIST_NAME, `venue=${venue} out of scope`);
  }

  const market = pkg?.market_state || {};
  const apr = Number(market.pool_apr_pct);
  const tvl = Number(market.tvl_usd);
  const arbSpread = Number(market.arb_spread_pct);
  const il = Number(market.impermanent_loss_pct);

  const citations = [{ type: 'signal', source: 'defi', venue, market }];

  if (!Number.isFinite(apr) && !Number.isFinite(arbSpread)) {
    return abstainBecause(DEFI_SPECIALIST_NAME, 'no pool/arb data');
  }

  // Open position with significant IL — sell.
  if (pkg?.trigger?.kind === 'sell_reeval' && Number.isFinite(il) && il > 8) {
    return voteFor(
      DEFI_SPECIALIST_NAME, 'sell', Math.min(0.85, 0.6 + il / 30),
      `impermanent loss ${il.toFixed(1)}% eating yield`,
      citations,
    );
  }

  // High APR + tvl = entry.
  if (apr > 50 && tvl > 1_000_000) {
    return voteFor(
      DEFI_SPECIALIST_NAME, 'buy', Math.min(0.82, 0.6 + apr / 300),
      `pool APR ${apr.toFixed(0)}% + TVL $${(tvl/1e6).toFixed(1)}M`,
      citations,
    );
  }

  // Arb opportunity = entry.
  if (arbSpread > 1.5) {
    return voteFor(
      DEFI_SPECIALIST_NAME, 'buy', Math.min(0.80, 0.55 + arbSpread / 5),
      `arb spread ${arbSpread.toFixed(2)}% available`,
      citations,
    );
  }

  return voteFor(
    DEFI_SPECIALIST_NAME, 'hold', 0.45,
    `APR ${Number.isFinite(apr) ? apr.toFixed(0)+'%' : 'n/a'} arb ${Number.isFinite(arbSpread) ? arbSpread.toFixed(2)+'%' : 'n/a'} — no edge`,
    citations,
  );
}

export const defiSpecialist = makeSpecialist(DEFI_SPECIALIST_NAME, ['defi'], argue);

File: go/deliberation/pkg/deliberation/specialists/defi/defi.go
// Package defi — Go port of @accretive-substrate/deliberation/specialists/defi.
//
// Filters venue in {aerodrome, uniswap}. Reads pool_apr_pct + tvl_usd +
// arb_spread_pct + impermanent_loss_pct.
package defi

import (
	"fmt"
	"math"
	"strings"

	"github.com/ramene/accretive-substrate/go/deliberation/pkg/deliberation"
)

const Name = "defi"

type defiMarket struct {
	PoolAPRPct           float64 `json:"pool_apr_pct"`
	TVLUSD               float64 `json:"tvl_usd"`
	ArbSpreadPct         float64 `json:"arb_spread_pct"`
	ImpermanentLossPct   float64 `json:"impermanent_loss_pct"`
}

func Argue(pkg deliberation.EvidencePackage) deliberation.Voice {
	venue := strings.ToLower(pkg.Trigger.Venue)
	if venue != "aerodrome" && venue != "uniswap" {
		return deliberation.AbstainBecause(Name, fmt.Sprintf("venue=%s out of scope", venue))
	}

	var market defiMarket
	_ = deliberation.DecodeMarketStateFor(pkg.MarketState, pkg.Trigger.Symbol, &market)

	citations := []deliberation.Citation{{
		Type:   "signal",
		Source: "defi",
	}}

	if market.PoolAPRPct == 0 && market.ArbSpreadPct == 0 {
		return deliberation.AbstainBecause(Name, "no pool/arb data")
	}

	// Open position with significant IL → sell.
	if pkg.Trigger.Kind == deliberation.TriggerSellReeval && market.ImpermanentLossPct > 8 {
		conf := math.Min(0.85, 0.6+market.ImpermanentLossPct/30)
		return deliberation.VoteFor(Name, deliberation.VerdictSell, conf,
			fmt.Sprintf("impermanent loss %.1f%% eating yield", market.ImpermanentLossPct),
			citations)
	}

	// High APR + TVL → buy.
	if market.PoolAPRPct > 50 && market.TVLUSD > 1_000_000 {
		conf := math.Min(0.82, 0.6+market.PoolAPRPct/300)
		return deliberation.VoteFor(Name, deliberation.VerdictBuy, conf,
			fmt.Sprintf("pool APR %.0f%% + TVL $%.1fM", market.PoolAPRPct, market.TVLUSD/1e6),
			citations)
	}

	// Arb opportunity → buy.
	if market.ArbSpreadPct > 1.5 {
		conf := math.Min(0.80, 0.55+market.ArbSpreadPct/5)
		return deliberation.VoteFor(Name, deliberation.VerdictBuy, conf,
			fmt.Sprintf("arb spread %.2f%% available", market.ArbSpreadPct),
			citations)
	}

	return deliberation.VoteFor(Name, deliberation.VerdictHold, 0.45,
		fmt.Sprintf("APR %.0f%% arb %.2f%% — no edge", market.PoolAPRPct, market.ArbSpreadPct),
		citations)
}

// Specialist — exported safe-wrapped argue function.
var Specialist = deliberation.MakeSafe(Name, Argue)
